File: src/main/java/org/minefortress/entity/ai/goal/WanderAroundTheFortressGoal.java
package org.minefortress.entity.ai.goal;

import net.minecraft.util.math.BlockPos;
import net.remmintan.mods.minefortress.core.interfaces.entities.pawns.controls.IEatControl;
import net.remmintan.mods.minefortress.core.interfaces.server.IServerFortressManager;
import net.remmintan.mods.minefortress.core.utils.ServerModUtils;
import org.minefortress.entity.Colonist;

import static org.minefortress.entity.colonist.FortressHungerManager.IDLE_EXHAUSTION;

public class WanderAroundTheFortressGoal extends AbstractFortressGoal {

    private BlockPos goal;

    public WanderAroundTheFortressGoal(Colonist colonist) {
        super(colonist);
    }

    @Override
    public boolean canStart() {
        if(colonist.getEatControl().map(IEatControl::isEating).orElse(false)) return false;
        if(!isDay() || colonist.getTaskControl().hasTask()) return false;

        // Using ServerModUtils.getFortressManager instead of colonist.getServerFortressManager()
        final var posOptional = ServerModUtils.getFortressManager(colonist)
                .flatMap(IServerFortressManager::getRandomPositionAroundCampfire);

        if(posOptional.isPresent()) {
            goal = posOptional.get();
            return true;
        }
        return false;
    }

    @Override
    public void start() {
        colonist.setCurrentTaskDesc("Wandering around");
        colonist.putItemInHand(null);
        colonist.getMovementHelper().goTo(goal, Colonist.SLOW_MOVEMENT_SPEED);
        if(colonist.isSleeping()) {
            colonist.wakeUp();
        }
    }

    @Override
    public void tick() {
        super.tick();
        colonist.addHunger(IDLE_EXHAUSTION);
        if(colonist.getMovementHelper().isStuck()) {
            if(goal != null) {
                colonist.resetControls();
                colonist.teleport(goal.getX(), goal.getY(), goal.getZ());
            }
        }
    }

    @Override
    public boolean shouldContinue() {
        return isDay() &&
                !colonist.getTaskControl().hasTask() &&
                (colonist.getMovementHelper().stillTryingToReachGoal() || colonist.getMovementHelper().isStuck());
    }

    @Override
    public void stop() {
        goal = null;
        colonist.getMovementHelper().reset();
    }

    private boolean isDay() {
        return colonist.getWorld().isDay();
    }
}

File: src/main/java/org/minefortress/entity/ai/goal/WanderAroundTheFortressGoal.kt
package org.minefortress.entity.ai.goal

import net.minecraft.util.math.BlockPos
import net.remmintan.mods.minefortress.core.utils.ServerModUtils
import org.minefortress.entity.Colonist
import org.minefortress.entity.colonist.IFortressHungerManager
import java.time.Duration
import java.time.LocalDateTime

class WanderAroundTheFortressGoal(colonist: Colonist) : AbstractFortressGoal(colonist) {
    private var goal: BlockPos? = null
    private var stopTime: LocalDateTime = LocalDateTime.now()
    private var delay: Int = 0

    override fun canStart(): Boolean {
        if (colonist.eatControl.map { it.isEating }.orElse(false)) return false
        if (!isDay || colonist.taskControl.hasTask()) return false
        val durationSinceStop = Duration.between(stopTime, LocalDateTime.now()).toMillis()
        if (durationSinceStop < delay) return false

        goal = ServerModUtils
            .getFortressManager(colonist)
            .map { it.randomFortressPosition }
            .orElse(null)

        return goal != null
    }

    override fun start() {
        colonist.currentTaskDesc = "Wandering around"
        colonist.putItemInHand(null)
        colonist.movementHelper.goTo(goal!!, Colonist.SLOW_MOVEMENT_SPEED)
    }

    override fun tick() {
        super.tick()
        colonist.addHunger(IFortressHungerManager.IDLE_EXHAUSTION)
        if (colonist.movementHelper.isStuck) {
            if (goal != null) {
                colonist.resetControls()
                colonist.teleport(goal!!.x.toDouble(), goal!!.y.toDouble(), goal!!.z.toDouble())
            }
        }
    }

    override fun shouldContinue(): Boolean {
        return isDay && !colonist.taskControl.hasTask() && colonist.movementHelper.stillTryingToReachGoal()
    }

    override fun stop() {
        goal = null
        colonist.movementHelper.reset()
        stopTime = LocalDateTime.now()
        delay = colonist.world.random.nextInt(6000) + 5000
    }

    private val isDay: Boolean
        get() = colonist.world.isDay
}
